### PR TITLE
Improvements to tests

### LINKED ISSUE #416: Check all tests for const-correctness and explanation
**Describe the current state of a part of the code**
Currently, I think the justification of a test is mostly only clear for @tmhuysen and @lelemmen.

**Describe a change that would improve the code base**
We should go over the tests and preferably add:
- const-correctness;
- documentation on top of the `BOOST_AUTO_TEST_CASE`, clearly indicating the reason for the test;
- documentation on the steps of the test case.

Here's an overview to indicate the progress:
- [x] LibintInterfacer_test;
- [x] IntegralCalculator_test;
- [x] GTOBasisSet_test;
- [x] GTOShell_test;
- [x] ScalarBasis_test;
- [x] ShellSet_test;
- [x] GSpinorBasis_test;
- [x] OrbitalSpace_test;
- [x] RSpinOrbitalBasis_test;
- [x] SimpleSpinorBasis_test;
- [x] USpinOrbitalBasis_test;
- [x] JacobiRotation_test;
- [x] OrbitalRotationGenerators_test;
- [x] SimpleTransformationMatrix_test;
- [x] Simple2DM_test;
- [x] SpinResolved1DM_test;
- [x] SpinResolved2DM_test;
- [x] CartesianExponents_test;
- [x] CartesianGTO_test;
- [x] ScalarFunction_test;
- [x] CubicGrid_test;
- [x] Field_test;
- [x] WeightedGrid_test;
- [x] Eigenpair_test;
- [x] EigenproblemSolver_test;
- [x] IterativeIdentitiesHessianModifier_test;
- [x] Minimizer_test;
- [x] NonLinearEquationSolver_test;;
- [x] DenseVectorizer_test;
- [x] ImplicitMatrixSlice_test;
- [x] ImplicitRankFourTensorSlice_test;
- [x] Matrix_test;
- [x] SquareMatrix_test;
- [x] SquareRankFourTensor_test;
- [x] Tensor_test;
- [x] elements_test;
- [x] Molecule_test;
- [x] NuclearFramework_test;
- [x] Nucleus_test;
- [x] ONVPath_test;
- [x] SeniorityZeroONVBasis_test;
- [x] SpinResolvedFrozenONVBasis_test;
- [x] SpinResolvedONV_test;
- [x] SpinResolvedONVBasis_test;
- [x] SpinResolvedSelectedONVBasis_test;
- [x] SpinUnresolvedFrozenONVBasis_test;
- [x] SpinUnresolvedONV_test;
- [x] SpinUnresolvedONVBasis_test;
- [x] NuclearDipoleOperator_test;
- [x] NuclearRepulsionOperator_test;
- [x] HoppingMatrix_test;
- [x] EvaluatableScalarRSQOneElectronOperator_test;
- [x] SimpleSQOneElectronOperator_test;
- [x] SimpleSQTwoElectronOperator_test;
- [x] SQHamiltonian_test;
- [x] USQOneElectronOperator_test;
- [x] USQTwoElectronOperator_test;
- [x] FrozenCoreDOCIRDMBuilder_test;
- [x] SpinResolvedFrozenDMCalculator_test;
- [x] expectation_values_test;
- [x] properties_test;
- [x] constrained_RHF_test;
- [x] QCMethod_CCD_test;
- [x] QCMethod_CCSD_test;
- [x] DOCI_test;
- [ ] DOCINewtonOrbitalOptimizer_test;
- [x] FCI_test;
- [x] Hubbard_test;
- [x] AP1roGJacobiOrbitalOptimizer_test;
- [x] AP1roGLagrangianNewtonOrbitalOptimizer_test;
- [x] QCMethod_AP1roG_test;
- [x] QCMethod_vAP1roG_test;
- [x] QCMethod_GHF_test;
- [x] QCMethod_RHF_test;
- [x] RHFSCFEnvironment_test;
- [x] QCMethod_UHF_test;
- [x] ERJacobiLocalizer_test;
- [x] ERNewtonLocalizer_test;
- [x] RMP2_test;
- [x] LinearExpansion_test;
- [x] AP1roGGeminalCoefficients_test;
- [x] APIGGeminalCoefficients_test;
- [x] RHF_test;
- [x] UHF_test;
- [x] linalg_test;
- [x] miscellaneous_test;
- [x] units_test;

### PR DESCRIPTION
**Pull request by lelemmen**
**Short description**

This PR aims to close #416.

### CODE REVIEW

File: gqcp/tests/Basis/Integrals/Interfaces/LibintInterfacer_test.cpp [hunk 1]
```diff
@@ -47,30 +47,34 @@ struct approx {
...
 
-    // Use the Libint interface to obtain a std::vector<libint2::Atom> from the GQCP ones
-    auto test_libint_atoms = GQCP::LibintInterfacer::get().interface(GQCP_nuclei);
+    // Use the Libint interface to obtain a std::vector<libint2::Atom> from the GQCP nuclei.
+    const auto test_libint_atoms = GQCP::LibintInterfacer::get().interface(GQCP_nuclei);
 
 
     /**
-     *  Implement a function object to compare (libint_atom) == (libint_atom)
+     *  A functor that compares compare (libint_atom) == (libint_atom)
```
Comment (dariatols) [inline gqcp/tests/Basis/Integrals/Interfaces/LibintInterfacer_test.cpp:77]:
```suggestion
     *  A functor that compares compare (libint_atom) == (libint_atom).
```

File: gqcp/tests/Basis/Transformations/OrbitalRotationGenerators_test.cpp [hunk 2]
```diff
@@ -41,18 +49,25 @@ BOOST_AUTO_TEST_CASE(asMatrix) {
...
+/**
+ *  Check if the `FromOccOcc` named constructor behaves correctly.
+ */
 BOOST_AUTO_TEST_CASE(FromOccOcc) {
 
+    // Initialize a test set of orbital rotation generators for an occupied-occupied orbital space.
     GQCP::VectorX<double> kappa {3};
     kappa << 1, 2, 3;
-    GQCP::OrbitalRotationGenerators occ_occ_generators {kappa};  // 3 occupied spatial orbitals
+    const GQCP::OrbitalRotationGenerators occ_occ_generators {kappa};  // 3 (doubly-)occupied spatial orbitals
```
Comment (dariatols) [inline gqcp/tests/Basis/Transformations/OrbitalRotationGenerators_test.cpp:67]:
```suggestion
    const GQCP::OrbitalRotationGenerators occ_occ_generators {kappa};  // 3 (doubly-)occupied spatial orbitals.
```

File: gqcp/tests/ONVBasis/SpinUnresolvedONVBasis_test.cpp [hunk 3]
```diff
@@ -448,7 +446,7 @@ BOOST_AUTO_TEST_CASE(iterateToNextUnoccupiedOrbital_signed) {
     //  Initial weight contributions were 1 and 1 respectively,
     //  these should be shifted to 2 and 3 respectively, the difference is 1 and 2 respectively.
     //  The total shift is thus 3, and the sign should remain the same (flips twice)
```
Comment (dariatols) [inline gqcp/tests/ONVBasis/SpinUnresolvedONVBasis_test.cpp:448]:
```suggestion
    //  The total shift is thus 3, and the sign should remain the same (flips twice).
```

File: gqcp/tests/ONVBasis/SpinUnresolvedONVBasis_test.cpp [hunk 4]
```diff
@@ -495,7 +493,7 @@ BOOST_AUTO_TEST_CASE(shiftToPreviousOrbital_signed) {
     //  Initial weight contribution is 1,
     //  This should be shifted to 0, the difference is 1.
     //  The sign changes (flips once)
```
Comment (dariatols) [inline gqcp/tests/ONVBasis/SpinUnresolvedONVBasis_test.cpp:495]:
```suggestion
    //  The sign changes (flips once).
```

File: gqcp/tests/ONVBasis/SpinUnresolvedONVBasis_test.cpp [hunk 5]
```diff
@@ -504,7 +502,7 @@ BOOST_AUTO_TEST_CASE(shiftToPreviousOrbital_signed) {
 
     sign = 1;
     address_shift = 0;
-    onv = fock_space.constructONVFromAddress(9);  // 11100
+    onv = onv_basis.constructONVFromAddress(9);  // 11100
     // test shift if we plan on creating two electrons and start from orbital index 2
```
Comment (dariatols) [inline gqcp/tests/ONVBasis/SpinUnresolvedONVBasis_test.cpp:506]:
```suggestion
    // Test shift if we plan on creating two electrons and start from orbital index 2.
```

File: gqcp/tests/ONVBasis/SpinUnresolvedONVBasis_test.cpp [hunk 6]
```diff
@@ -514,7 +512,7 @@ BOOST_AUTO_TEST_CASE(shiftToPreviousOrbital_signed) {
     //  Initial weight contribution is 2,
     //  This should be shifted to 0, the difference is 2.
     //  The sign changes (flips once)
```
Comment (dariatols) [inline gqcp/tests/ONVBasis/SpinUnresolvedONVBasis_test.cpp:514]:
```suggestion
    //  The sign changes (flips once).
```

**Review by dariatols — APPROVED** [review]
This PR looks good to go, thanks for the refinements that make the tests easier to read. I have added some minor adjustments on lines where no punctuation were used.

**Review by xdvriend — APPROVED** [review]
Good to go for me!

**Pull request merged**